Media search indexing (Lombiq Technologies: OCORE-55)

Piedone opened the pull request:
Provides a simple implementation to index media for search. More precisely, it indexes media files related to content items so said content items will turn up in search when one of their media files matches the search query.

The following data can be indexed for each file referenced from a Media Field:

- Media Text
- Textual content of PDF files

agriffard commented:
Please add the license and version of https://github.com/UglyToad/PdfPig to https://docs.orchardcore.net/en/dev/docs/resources/libraries/

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Media/Handlers/MediaFieldIndexHandler.cs:
```diff
@@ -0,0 +1,62 @@
+                        {
+                            foreach (var key in context.Keys)
+                            {
+                                context.DocumentIndex.Set(key, page.Text, options);
```
This should always be analyzed.

Piedone replied:
Why? We have the same way of configuring indexing for all fields, so whether to analyze and/or store is coming from the field settings.

Skrypt replied:
Is it relevant to index/store the entire PDF page to search for a term in it?

Piedone replied:
I don't see why it wouldn't be.

Skrypt replied:
key + ".PdfText"

Piedone replied:
While I changed the key I'm still not sure about your initial remark.

Skrypt replied:
The only advantage I can see is if we allow to "store" the text of a PDF it won't need to be parsed again by PdfPig. Though, what you want is to be able to find specific terms in these PDF so, you want to analyze/tokenize that text more than just storing it. Storing is generally used on technical values like a GUID or a SKU. These makes sense to be only stored as they are generally unique.

Piedone replied:
Hmm, I see. I could hard-code `options |= DocumentIndexOptions.Analyze;` here but wouldn't it be awkward that it doesn't fully honor the configuration selected on the admin?

Skrypt replied:
It would still honor the configuration for the MediaText. We do this also for the TitlePart ... I think what we eventually will need is something to disable these configs per field if needed. Here for example we would need 2 different configs since we index 2 different values. These indexes configurations needs to be enhanced to allow doing this kind of thing. It's been thinked only for fields that indexes a single value.

Piedone replied:
OK, then, added it.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Media/Handlers/MediaFieldIndexHandler.cs:
```diff
@@ -0,0 +1,62 @@
+                    {
+                        foreach (var mediaText in field.MediaTexts)
+                        {
+                            context.DocumentIndex.Set(key, mediaText, options);
```
This can be stored or analyzed

Skrypt replied:
key + ".MediaText"

Piedone commented:
Guys, slow down :D.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media/Handlers/MediaFieldIndexHandler.cs:
```diff
@@ -0,0 +1,62 @@
+                    }
+                }
+
+                foreach (var path in field.Paths.Where(path => path.EndsWith(".pdf")))
```
Sebastien's idea: Make this extensible, just via options (not DI).

Piedone replied:
I just recognized I don't know what you meant by options @sebastienros. If `IOptions<T>` then I'm not sure how providing an extension point for services that would hook into this indexing work. I'd rather create an injectable `IMediaFieldFileIndexHandler` and run through them here.

deanmarcussen replied:
@Piedone like the feature, it is the thing that I always recomend to clients not do do, is use too many pdfs, because indexing and seo is dificult with them.

re options: using `Options<T>` vs providers is essentially the same thing, just potentially easier to configure, and less expensive, in terms of DI lookups etc. There are probably more places we could have used the Options pattern, over a DI pattern.

Consider for example this could just be a dictionary of funcs on an options class.

```
public Dictionary<string, Func<Stream, string[]>> MediaPaths;
```

then in the `MediaFieldIndexHandler`

```
if (_options.TryGetValue(fileExtension, out var indexer)
{
   var values = indexer(fileStrem);
}
```

making it trivial to `Configure` a `.txt` file indexer etc al.

Of course exactly the same thing could be done with a provider, so as you prefer.

In terms of the `FullText` index, there are two ways to go here, as you have done, by adding the new fields to the Search Index, or using the `Custom Full Text Settings` liquid editor on a content type, to return these values into the `FullTextIndex`. I tend towards doing that more, as it makes it easier to differentiate between indexes that we might use internally for custom lucene queries, and not have to add so many fields to the index. For example with the `full_text` liquid aspect, we are also able to index into `Widgets` and `BagPart` items, and control which fields on these contained items are used for the full text index, from their own content definitions.

For example media text can already be added to a full text index, simply by returning (in a loop) `{{ Content.MyPart.MyField.MediaTexts }}`

What we also need at some point, is the same feature as the Full Text Index Settings for the `BodyAspect` to allow Widgets and BagParts to better contribute to a `BodyAspect` for rss feeds etc.

Whatever you decide in terms of providers vs options, it would be good, if it was reusable (later or now) in a liquid filter, to achieve the same thing as we can with `| full_text` 

and what @Skrypt says about about the key names

Piedone replied:
Ah, thanks for the clarification on options. I'd prefer the provider-based approach though: While if the configured file content extractor does nothing apart from processing the stream having such delegates in an options class is easier, if it needs to access injected services too it'd become cumbersome for the consumer (or we'd need to pass and `IServiceProvider` around as `IBackgroundTask` does).

As for how to handle FullText:

While handling `MediaText` can be done as follows too:

```liquid
{% for mediaText in Model.Content.ContentTypeName.FieldName.MediaTexts %}
  {{ mediaText }}
{% endfor %}
```

I think doing something similar for file contents would complicate things quite a lot (both on the implementation and configuration side), and be an exception to what fields usually do (which is saving fields in a `ContentFieldIndexHandler<T>`.

In the wider scope I'd probably prefer all fields that are configured to be indexed (and stored and analyzed) to be automatically added to `Content.ContentItem.FullText` as well, so no further configuration is necessary to make it searchable too. So maybe a `FullTextAspect`-populating handler for each content field?

Skrypt replied:
I see your point and it kind of makes sense but not at all times. So, I guess we should have some global recipes for Content Types FullText indexing. This would be a global full-text configuration that would be overriden by the `Content Type` settings themselves. I'm not sure if it's a good idea to put everything in the full-text column while you can already search in every field with a Lucene Query. I kind of understand that since every content types have different fields it makes it harder to keep all the Content Types full-text settings to be properly configured when people are adding new fields which should be added to the full-text index. Though, it is wrong to assume that every field should be part of it. So this would require some more thinking.

Piedone replied:
Perhaps a field setting to add it to `FullText` optionally? Thus have a simple way to use them for search but keep the option open for more involved usage in custom queries.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media/Indexing/PdfMediaFileTextProvider.cs:
```diff
@@ -0,0 +1,32 @@
+            return path.EndsWith(".pdf");
+        }
+
+        public string GetText(string path, Stream fileStream)
```
I used the approach to return `string` here so later perhaps the same providers can be used to build `Content.ContentItem.FullText` too. However, it also means that we'll pass around potentially large strings.

The alternative though would be to let these providers do the indexing directly (i.e. `context.DocumentIndex.Set()`) so not the whole file content needs to be loaded at once. That wouldn't keep the option open to use them with `Content.ContentItem.FullText` too though.

What do you think, @deanmarcussen?

Piedone replied:
What do you think, @deanmarcussen?

Skrypt replied:
I think we should keep the FullText Aspect separated with the FileText Aspect. Here, if we want to search in both we can still do it with Lucene. But for making this cleaner I'd not mix those 2 aspects.

Piedone replied:
If there will be a `FileText` aspect then I guess it would need to receive the string content of the file in the same way then. Doing something with aspects is an option for the future though.

Skrypt replied:
Yeah, we don't have a FileText aspect and probably won't need. But at the same time if someone wants to add the FileText content in his FullText index it will kind of duplicate indexing for nothing. Here, if we already analyze the FileText document in the index then we should just use it instead of adding it also in the FullText document. It will make the index grow unecessarly. That's more what I meant. Also, in the Search feature we can specify which index column(s) we want to look into. So, it's just a matter of adding the FileText column to the Search configuration.

Skrypt replied:
Also, I would tend to create a completely independent search controller for searching in these PDF files. I generally want to search inside the website page content independently of their attached media files. An example could be a search feature for the Media Gallery that would allow to search inside any files text content.

Piedone replied:
For your first message: Since this is what happens now do I understand correctly that what's now in the code is good then?

For the second one: We could spec out a file search feature for Media Library as a next step, the infrastructure for it will be here then.

Skrypt replied:
Yes, all good.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media/Indexing/MediaFieldIndexHandler.cs:
```diff
@@ -0,0 +1,77 @@
+                        using var fileStream = await _mediaFileStore.GetFileStreamAsync(path);
+                        if (fileStream != null)
+                        {
+                            var fileTexts = _mediaFileTextProviders
+                                .Where(provider => provider.CanHandle(path))
+                                .Select(provider => provider.GetText(path, fileStream));
```
Here multiple providers can produce their own versions of the given file's text. Let's suppose these are different. Do you think @Skrypt it would make sense to index all such versions? An alternative is to pick the last (i.e. lastly registered) matching provider and produce the text with that only.

Piedone replied:
What do you think, @Skrypt?

Skrypt replied:
I assume it would produce multiple versions if we had 2 different providers that CanHandle a .pdf file for example. Then, here I think it would be better to index only one version (lastly registered). So basically, this would allow to override the providers to use a different PDF parsing engine.

Piedone replied:
Thanks!

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Media/Indexing/PdfMediaFileTextProvider.cs:
```diff
@@ -0,0 +1,32 @@
+    {
+        public bool CanHandle(string path)
+        {
+            return path.EndsWith(".pdf");
```
Case insensitive probably?

Piedone replied:
Good point.

deanmarcussen reviewed (commented):
Can you take it out of draft @Piedone ?

Piedone commented:
@deanmarcussen it is a draft though, not finished at all.

sebastienros commented:
My suggestion is to use an Options class, not DI, such that one filter can be registered per extension (dictionary). This allows flexibility to replace or override existing filters. And finding the filter for an extension is also more straightforward. Can also list then in an options screen (something @agriffard has done for some features to "understand" what's used by the system).

Example we have moved to this pattern: Content Parts and Fields, Liquid template options.

hishamco commented on src/docs/resources/libraries/README.md:
```diff
@@ -28,6 +28,7 @@ The below table lists the different .NET libraries used in Orchard Core:
 | [Newtonsoft.Json](https://github.com/JamesNK/Newtonsoft.Json) | Json.NET is a popular high-performance JSON framework for .NET | 13.0.1 | [MIT](https://github.com/JamesNK/Newtonsoft.Json/blob/master/LICENSE.md) |
 | [Noda Time](https://github.com/nodatime/nodatime) | A better date and time API for .NET. | 3.0.9 | [Apache-2.0](https://github.com/nodatime/nodatime/blob/master/LICENSE.txt) |
 | [OpenIddict](https://github.com/openiddict/openiddict-core) | Versatile OpenID Connect stack for ASP.NET Core. | 3.1.1 | [Apache-2.0](https://github.com/openiddict/openiddict-core/blob/dev/LICENSE.md)) |
+| [PdfPig](https://github.com/UglyToad/PdfPig/) | Library to read and extract text and other content from PDF files. | 0.1.5-alpha001 | [Apache-2.0](https://github.com/UglyToad/PdfPig/blob/master/LICENSE) |
```
It's no longer alpha https://www.nuget.org/packages/PdfPig/

Piedone replied:
Indeed, thanks, updated.

hishamco commented on src/docs/reference/modules/Media/README.md:
```diff
@@ -475,8 +475,35 @@ When the query string is signed with a token any width, height value may be used
 
+## Media Indexing
+
+Media can be optionally indexed for search as well if files are referenced via Media Fields. The following data can be indexed for each file referenced from a Media Field:
```
Can we add apostrophe in `Media Field`?

Piedone replied:
What do you mean? I don't see a need for an apostrophe here ("files are referenced via Media Fields" is in plural and shouldn't be "files are referenced via Media Field's").

hishamco replied:
It's not the matter of plural, I just want markdown highlight the text ;)

hishamco replied:
It's not the matter of plural, I just want markdown highlight the text ;)

Piedone replied:
I really don't understand what you mean. This?

```suggestion
Media can be optionally indexed for search as well if files are referenced via `Media Field`s. The following data can be indexed for each file referenced from a `Media Field`:
```

However, these are not technical identifiers but English nouns. What I'd denote with backticks would be "MediaField", as in the name of the class.

hishamco commented on src/OrchardCore.Modules/OrchardCore.Media/Indexing/PdfMediaFileTextProvider.cs:
```diff
@@ -0,0 +1,32 @@
+    {
+        public bool CanHandle(string path)
+        {
+            return path.EndsWith(".pdf", System.StringComparison.OrdinalIgnoreCase);
```
```csharp
private static readonly string PdfExtension = ".PDF";

public bool CanHandle(string path) => path.EndsWith(PdfExtension, System.StringComparison.OrdinalIgnoreCase);
```

Piedone replied:
Why would it be useful to pull out a single-use string into a private field?

As for having an expression-bodied method, I actually prefer that personally but I think the practice in Orchard is not to do it (unlike for properties).

Skrypt replied:
This one I won't debate as it is mostly a preference than anything. No value added.

hishamco commented on src/OrchardCore.Build/Dependencies.props:
```diff
@@ -39,6 +39,7 @@
     <PackageManagement Include="OpenIddict.AspNetCore" Version="3.1.1" />
     <PackageManagement Include="OpenIddict.Core" Version="3.1.1" />
     <PackageManagement Include="OrchardCore.Translations.All" Version="1.1.0" />
+    <PackageManagement Include="PdfPig" Version="0.1.5-alpha001" />
```
It's no longer alpha https://www.nuget.org/packages/PdfPig/

Piedone replied:
Thanks, updated here too.

hishamco commented on src/OrchardCore.Modules/OrchardCore.Media/Indexing/IMediaFileTextProvider.cs:
```diff
@@ -0,0 +1,22 @@
+        /// </remarks>
+        bool CanHandle(string path);
+
+        string GetText(string path, Stream fileStream);
```
If we have a `path` we can get the stream instead of passing it

Piedone replied:
Not necessarily. Path here can be any identifier of the file, including a Blob Storage URL even. This is only needed to be able to determine whether it's a file type we can handle, by looking at the extension (this is not foolproof by any means but the quickest and easiest way to do it). Hence we also require a `Stream` to get the content of the file, whatever form it's stored.

Skrypt replied:
I kind of agree with @hishamco on that one but it's not a big deal.

```C#
    string extension = Path.GetExtension(myFileStream.Name);
```

hishamco commented on src/OrchardCore.Modules/OrchardCore.Media/Indexing/IMediaFileTextProvider.cs:
```diff
@@ -0,0 +1,22 @@
+        /// can simply mean the file's textual content is empty.
+        /// </para>
+        /// </remarks>
+        bool CanHandle(string path);
```
This could be a utility or `private` method that's utilized by `GetText()`, but I will leave others to comment

hishamco replied:
Seems it's used to filter `IMediaFileTextProvider`s, right?

Piedone replied:
Yes, and ;):

![image](https://user-images.githubusercontent.com/1976647/141661411-c0180a80-cc10-41ee-b8d5-b71a6649f7e3.png)

hishamco commented on src/OrchardCore.Modules/OrchardCore.Media/Indexing/MediaFieldIndexHandler.cs:
```diff
@@ -0,0 +1,77 @@
+            {
+                foreach (var key in context.Keys)
+                {
+                    context.DocumentIndex.Set(key + ".MediaText", "NULL", options);
```
Can we introduce static fields for both ".MediaText" and ".FileText"?

Piedone replied:
Sure, done that.

Piedone commented:
I addressed all the feedback. Extensibility is still based on injected interfaces, not options. See the rationale here: https://github.com/OrchardCMS/OrchardCore/pull/9544#discussion_r642136433

Please review when you can.

Skrypt commented:
@deanmarcussen please review before we merge.

hishamco commented:
It would be nice to complete the docs that you already started, or I will do it in another PR

deanmarcussen reviewed (changes requested):
> My suggestion is to use an Options class, not DI, such that one filter can be registered per extension (dictionary). This allows flexibility to replace or override existing filters. And finding the filter for an extension is also more straightforward. Can also list then in an options screen (something @agriffard has done for some features to "understand" what's used by the system).
> 
> Example we have moved to this pattern: Content Parts and Fields, Liquid template options.

Both @sebastienros and I have suggested an options class for this. It makes much more sense based on how we are trying to do things now - allows much more flexibility to configure easily, and using DI for providers like this is really bad for performance (because the container instantiates everything, even if 9 times out of ten is is not necesary)

The concern about having to pass a service provider is easily resolvable.
Here are some examples of how this can be achieved.

https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.Workflows.Abstractions/Helpers/ServiceCollectionExtensions.cs
https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.Workflows.Abstractions/Helpers/ServiceProviderExtensions.cs

I would imagine an Options Dictionary here could be keyed by extension name, and contain a `Type`.

The `MediaFieldIndexHandler` then just injects the options dictionary, a service provider, and instantiates the type of handler as required.

This has a lot of advantages, because the options are able to easily add/remove configurations, the text `Type` becomes built as and when required, we lose the `CanHandle`

Skrypt commented:
Ok, then @deanmarcussen if you can do these changes that would go faster. Else, we wait on these changes.

Piedone commented:
The docs are complete as far as I'm concerned @hishamco, what do you mean?

Well, @deanmarcussen you let me decide and Sebastien most possibly didn't read my prior explanation, so... :) But fine, I'll change it to options.

hishamco commented:
I mean code docs ;)

Piedone commented:
Implemented provider registration with options.

Piedone commented:
@deanmarcussen would you please check? Apart from your approval this is good to go.

deanmarcussen commented on src/OrchardCore/OrchardCore.Media.Abstractions/Indexing/MediaFileIndexingOptions.cs:
```diff
@@ -0,0 +1,50 @@
+{
+    public class MediaFileIndexingOptions
+    {
+        private readonly Dictionary<string, Type> _mediaFileTextProviderRegistrations = new Dictionary<string, Type>();
```
```suggestion
        private readonly Dictionary<string, Type> _mediaFileTextProviderRegistrations = new Dictionary<string, Type>(StringComparison.OrdinalIgnoreCase);
```

deanmarcussen commented on src/OrchardCore/OrchardCore.Media.Abstractions/Indexing/MediaFileIndexingOptions.cs:
```diff
@@ -0,0 +1,50 @@
+            return _mediaFileTextProviderRegistrations.Any();
+        }
+
+        private static string NormalizeFileExtension(string fileExtension)
```
This can go, if you use ordinal ignore case on the dictionary.

Or you could keep it for the `.` exception, and just remove the `ToUpperInvariant()`

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Media/Indexing/MediaFieldIndexHandler.cs:
```diff
@@ -0,0 +1,88 @@
+                    foreach (var path in field.Paths)
+                    {
+                        var extensionWithoutDot = Path.GetExtension(path).Substring(1);
+                        var providerType = _mediaFileIndexingOptions.GetRegisteredMediaFileTextProvider(extensionWithoutDot);
```
if you just prefixed a `.` to the registration methods, i.e. something like `var extensions = ext.TrimStart('.') + '.' then it would be normalized to  being with, and no need to allocate another string here, just for the `SubString()`

Piedone replied:
I think the `AddMediaFileTextProvider()` is more convenient if you don't have to always specify the dot. But probably this is worthwhile.

deanmarcussen commented on src/OrchardCore/OrchardCore.Media.Abstractions/Indexing/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,28 @@
+        /// The file extension of those files, without a leading dot, that the given <see
+        /// cref="IMediaFileTextProvider"/> implementation will process.
+        /// </param>
+        public static void AddMediaFileTextProvider<TMediaFileTextProvider>(this IServiceCollection services, string fileExtension)
```
```suggestion
        public static IServiceCollection AddMediaFileTextProvider<TMediaFileTextProvider>(this IServiceCollection services, string fileExtension)
```

deanmarcussen commented on src/OrchardCore/OrchardCore.Media.Abstractions/Indexing/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,28 @@
+        public static void AddMediaFileTextProvider<TMediaFileTextProvider>(this IServiceCollection services, string fileExtension)
+            where TMediaFileTextProvider : class, IMediaFileTextProvider
+        {
+            services.Configure<MediaFileIndexingOptions>(options => options.RegisterMediaFileTextProvider<TMediaFileTextProvider>(fileExtension));
```
```suggestion
            return services.Configure<MediaFileIndexingOptions>(options => options.RegisterMediaFileTextProvider<TMediaFileTextProvider>(fileExtension));
```

or just `=>` if you prefer

deanmarcussen commented on src/OrchardCore/OrchardCore.Media.Abstractions/Indexing/MediaFileIndexingOptions.cs:
```diff
@@ -0,0 +1,50 @@
+            return null;
+        }
+
+        public bool AnyMediaFileTextProviders()
```
What we normally do is just expose the 
`private readonly Dictionary<string, Type> _mediaFileTextProviderRegistrations = new Dictionary<string, Type>();`

as a readonly dictionary, so we don't need all these extra methods, for `Any()` etc.

so

`public IReadOnlyDictionary<string, Type> MediaFileTextProviders => _mediaFileTextProviders`

but up to you

Piedone replied:
I'd hide that implementation detail.

deanmarcussen reviewed (commented):
Yeah looks good, just a couple of tweaks

Piedone commented:
Implemented the suggestions.

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Media/Indexing/MediaFieldIndexHandler.cs:
```diff
@@ -0,0 +1,87 @@
+                    }
+                }
+
+                if (_mediaFileIndexingOptions.AnyMediaFileTextProviders())
```
This check is actually unecesary, because this is not a feature, so pdf indexing is always enabled, so there will always be a registered pdf indexer.

Unless you implement a `Remove` method (I don't see that we need it, so I'd just drop the `.Any()` check)

Piedone replied:
Indeed, though defensive checks like this are useful if the feature is changed later. Anyway, removed it.

deanmarcussen commented on src/OrchardCore/OrchardCore.Media.Abstractions/Indexing/MediaFileIndexingOptions.cs:
```diff
@@ -0,0 +1,51 @@
+
+        public MediaFileIndexingOptions RegisterMediaFileTextProvider(string fileExtension, Type providerType)
+        {
+            // Deliberate overwrite behavior so you can override registrations.
```
The suggestion here was to add
`var extension = fileExtension.TrimStart('.') + '.'` during registration so that the `.` is always prefixed (this is a lazy singleton, so will only happen once), and it has been normalized.

At which point you can remove the `NormalizeFileExtension` method, because no need to check it? as we know that `Path.GetExtension` includes the `.` right?

Otherwise `NormalizeFileExtension` should prepend the `.` rather than throwing an exception, or it needs a different name, because it's not normalizing ;)

I'd just drop it myself - keep the whole thing simpler

Piedone replied:
This is a published part of the interface that any developer may use, just as `GetRegisteredMediaFileTextProvider()` (though that perhaps less so, I still see the utility in keeping it open). Thus, we need validation. And while as I mentioned having to always include a dot in `AddMediaFileTextProvider()` and `RegisterMediaFileTextProvider()` (something that other developers will use) is inconvenient, perhaps it's acceptable to reduce allocations, as you asked for: Because if we require it in `GetRegisteredMediaFileTextProvider() then it would be confusing to not require the leading dot in these. Thus while `NormalizeFileExtension()` is now indeed incorrectly named (fixed it) if it would prepend the dot then we're back at having additional allocations, or an inconsistent API. I find both of these unfavorable compared to having the current check.
